src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,190 @@
+	Pos.x = Clamp(Pos.x, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+	Pos.z = Clamp(Pos.z, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+
+	// Place obsidian pillar
+	for (int Y = 0; Y <= Properties.m_Height - 2; Y++)
+	{
+		for (int X = -Properties.m_Radius; X < Properties.m_Radius; X++)
+		{
+			for (int Z = -Properties.m_Radius; Z < Properties.m_Radius; Z++)
+			{
+				Vector3i NewPos = {Pos.x + X, Y, Pos.z + Z};
+				// The 3 was achieved by trial and error till the shape matched the notchian implementation
+				if ((NewPos - Vector3i(Pos.x, Y, Pos.z)).SqrLength() < Properties.m_Radius * Properties.m_Radius - 3)
+				{
+					a_ChunkDesc.SetBlockType(NewPos.x, NewPos.y, NewPos.z, E_BLOCK_OBSIDIAN);
+				}
+			}
+		}
+	}
+	// Spawn the bedrock
+	a_ChunkDesc.SetBlockType(Pos.x, Properties.m_Height - 1, Pos.z, E_BLOCK_BEDROCK);
+	// Spawn the Ender Crystal - the original passed position is not used because the tower might have been moved due to crossing chunk borders
+	auto EnderCrystal = std::make_unique<cEnderCrystal>(Vector3d(Pos.x + cChunkDef::Width * a_ChunkDesc.GetChunkX() + 0.5, Properties.m_Height, Pos.z + cChunkDef::Width * a_ChunkDesc.GetChunkZ() + 0.5));
+	a_ChunkDesc.GetEntities().emplace_back(std::move(EnderCrystal));
+	// Spawn the fire
+	a_ChunkDesc.SetBlockType(Pos.x, Properties.m_Height, Pos.z, E_BLOCK_FIRE);
+	// Spawn the iron bars if there are some
+	if (Properties.m_HasCage)
+	{
+		// Place walls
+		for (int I = -2; I <= 2; I++)
+		{
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height - 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height + 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height - 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height + 1, Pos.z + I, E_BLOCK_IRON_BARS);
+
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height - 1, Pos.z - 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height, Pos.z - 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height + 1, Pos.z - 2, E_BLOCK_IRON_BARS);
+
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height - 1, Pos.z + 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height, Pos.z + 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height + 1, Pos.z + 2, E_BLOCK_IRON_BARS);
+		}
+		// Place roof
+		for (int X = -2; X <= 2; ++X)
+		{
+			for (int Z = -2; Z <= 2; ++Z)
+			{
+				a_ChunkDesc.SetBlockType(Pos.x + X, Properties.m_Height + 2, Pos.z + Z, E_BLOCK_IRON_BARS);
+			}
+		}
+	}
+	// Set the height to zero that this tower has been generated
+	m_TowerProperties[Index].m_Height = 0;
```

Comment:
Any reason not to remove the tower from the list entirely?
```suggestion
	std::swap(m_TowerProperties[Index], m_TowerProperties.back());
	m_TowerProperties.resize(m_TowerProperties.size() - 1);
```

Reply:
Yeah - there might be some tinkering to do if we want to place them 100% right